File: db4on/tutorial/db4o-tutorial-chapters/src/com/db4o/f1/chapter6/EvaluationExample.cs
using System.IO;
using com.db4o.f1.chapter3;
using com.db4o.query;

namespace com.db4o.f1.chapter6
{
	public class EvaluationExample : Util
	{
		public static void Main(string[] args)
		{
			File.Delete(Util.YapFileName);
			ObjectContainer db = Db4o.OpenFile(Util.YapFileName);
			try
			{
				StoreCars(db);
				QueryWithEvaluation(db);
			}
			finally
			{
				db.Close();
			}
		}

		public static void StoreCars(ObjectContainer db)
		{
			Pilot pilot1 = new Pilot("Michael Schumacher", 100);
			Car car1 = new Car("Ferrari");
			car1.Pilot = pilot1;
			car1.Snapshot();
			db.Set(car1);
			Pilot pilot2 = new Pilot("Rubens Barrichello", 99);
			Car car2 = new Car("BMW");
			car2.Pilot = pilot2;
			car2.Snapshot();
			car2.Snapshot();
			db.Set(car2);
		}

		public static void QueryWithEvaluation(ObjectContainer db)
		{
			Query query = db.Query();
			query.Constrain(typeof (Car));
			query.Constrain(new EvenHistoryEvaluation());
			ObjectSet result = query.Execute();
			Util.ListResult(result);
		}
	}
}

File: db4on/tutorial/vb/db4o-tutorial-chapters/src/com/db4o/f1/chapter6/EvaluationExample.vb
Imports System.IO
Imports com.db4o.query
Imports com.db4o

Namespace com.db4o.f1.chapter6
	Public Class EvaluationExample
	Inherits Util
		Public Shared Sub Main(ByVal args As String())
			File.Delete(Util.YapFileName)
            Dim db As Global.com.db4o.ObjectContainer = Db4oFactory.OpenFile(Util.YapFileName)
			Try
				StoreCars(db)
				QueryWithEvaluation(db)
			Finally
				db.Close()
			End Try
		End Sub

		Public Shared Sub StoreCars(ByVal db As ObjectContainer)
            Dim pilot1 As chapter3.Pilot = New chapter3.Pilot("Michael Schumacher", 100)
            Dim car1 As chapter3.Car = New chapter3.Car("Ferrari")
			car1.Pilot = pilot1
			car1.Snapshot()
			db.[Set](car1)
            Dim pilot2 As chapter3.Pilot = New chapter3.Pilot("Rubens Barrichello", 99)
            Dim car2 As chapter3.Car = New chapter3.Car("BMW")
			car2.Pilot = pilot2
			car2.Snapshot()
			car2.Snapshot()
			db.[Set](car2)
		End Sub

		Public Shared Sub QueryWithEvaluation(ByVal db As ObjectContainer)
			Dim query As Query = db.Query()
            query.Constrain(GetType(chapter3.Car))
			query.Constrain(New EvenHistoryEvaluation())
			Dim result As ObjectSet = query.Execute()
			Util.ListResult(result)
		End Sub

	End Class
End Namespace
